File: java/hangman/src/main/java/Hangman.java
import io.reactivex.Observable;
import io.reactivex.subjects.BehaviorSubject;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class Hangman {

    Observable<Output> play(Observable<String> words, Observable<String> letters) {
        // Create a BehaviorSubject to hold the current state of the game
        BehaviorSubject<Output> gameState = BehaviorSubject.createDefault(Output.empty());
        
        // Combine the words and letters observables to create the game logic
        return words
            // For each word, initialize a new game state
            .switchMap(word -> {
                // Create initial state for this word
                String maskedWord = "_".repeat(word.length());
                Output initialState = new Output(
                    word,
                    maskedWord,
                    new HashSet<>(),
                    new HashSet<>(),
                    new ArrayList<>(),
                    Status.PLAYING
                );
                
                // Update the game state with the initial state
                gameState.onNext(initialState);
                
                // Process each letter guess
                return letters
                    .scan(initialState, (state, letter) -> {
                        // Check if the letter has already been played
                        if (state.guess.contains(letter) || state.misses.contains(letter)) {
                            throw new IllegalArgumentException("Letter " + letter + " was already played");
                        }
                        
                        // Check if the game is already over
                        if (state.status != Status.PLAYING) {
                            return state;
                        }
                        
                        // Check if the letter is in the word
                        boolean isCorrectGuess = word.contains(letter);
                        
                        // Create new sets for guesses and misses
                        Set<String> newGuesses = new HashSet<>(state.guess);
                        Set<String> newMisses = new HashSet<>(state.misses);
                        List<Part> newParts = new ArrayList<>(state.parts);
                        
                        // Update the discovered word
                        StringBuilder newDiscovered = new StringBuilder(state.discovered);
                        
                        if (isCorrectGuess) {
                            // Add to guesses
                            newGuesses.add(letter);
                            
                            // Update the discovered word
                            for (int i = 0; i < word.length(); i++) {
                                if (word.charAt(i) == letter.charAt(0)) {
                                    newDiscovered.setCharAt(i, letter.charAt(0));
                                }
                            }
                        } else {
                            // Add to misses
                            newMisses.add(letter);
                            
                            // Add a new part to the hangman
                            if (newParts.size() < Part.values().length) {
                                newParts.add(Part.values()[newParts.size()]);
                            }
                        }
                        
                        // Check if the game is won or lost
                        Status newStatus = state.status;
                        if (!newDiscovered.toString().contains("_")) {
                            newStatus = Status.WIN;
                        } else if (newParts.size() >= Part.values().length) {
                            newStatus = Status.LOSS;
                        }
                        
                        // Create and return the new state
                        return new Output(
                            word,
                            newDiscovered.toString(),
                            newGuesses,
                            newMisses,
                            newParts,
                            newStatus
                        );
                    })
                    // Emit each state update
                    .doOnNext(gameState::onNext);
            })
            // Return the game state observable
            .startWith(gameState)
            .flatMap(output -> gameState);
    }
}

File: python/hangman/hangman.py
# Game status categories
# Change the values as you see fit
STATUS_WIN = 'win'
STATUS_LOSE = 'lose'
STATUS_ONGOING = 'ongoing'


class Hangman:
    def __init__(self, word):
        self.word = word
        self.remaining_guesses = 9
        self.status = STATUS_ONGOING
        self.guessed_letters = set()

    def guess(self, char):
        if self.status != STATUS_ONGOING:
            raise ValueError("The game has already ended.")
        
        if char in self.guessed_letters:
            self.remaining_guesses -= 1
        else:
            self.guessed_letters.add(char)
            if char not in self.word:
                self.remaining_guesses -= 1
        
        # Check if the game is lost
        if self.remaining_guesses < 0:
            self.status = STATUS_LOSE
        
        # Check if the game is won
        if all(letter in self.guessed_letters for letter in self.word):
            self.status = STATUS_WIN
            
    def get_masked_word(self):
        return ''.join(letter if letter in self.guessed_letters else '_' for letter in self.word)

    def get_status(self):
        return self.status
